elasticdl/python/master/task_manager.py

```diff
@@ -448,5 +463,67 @@ def set_evaluation_service(self, evaluation_service):
                 evaluation_service.init_eval_only_job(len(self._eval_todo))
 
     def start(self):
-        # TODO: start a thread to monitor the task execution time
-        pass
+        threading.Thread(
+            target=self._check_and_reassign_timeout_tasks,
+            name="check_timeout_tasks",
+            daemon=True,
+        ).start()
+
+    def reset_worker_start_task_time(self, worker_id):
+        self._worker_start_task_time[worker_id] = time.time()
+
+    def record_task_completed_time(self, task_type, completed_time):
+        self._task_completed_times[task_type].append(completed_time)
+
+    def set_task_timeout_callback(self, callback_fn):
+        self._task_timeout_callbacks.append(callback_fn)
+
+    def _invoke_task_timeout_callback(self, worker_id):
+        for callback_fn in self._task_timeout_callbacks:
+            callback_fn(worker_id)
+
+    def _check_and_reassign_timeout_tasks(self):
+        """Check whether there are timeout tasks periodically.
+        """
+        while True:
+            doing_tasks = self._doing.copy()
+            cur_time = time.time()
+            avg_time = self._get_average_task_completed_time()
+            for _, (worker_id, task, start_time) in doing_tasks.items():
+                if task.type == elasticdl_pb2.TRAINING:
+                    start_time = self._worker_start_task_time[worker_id]
+                if task.type in [
+                    elasticdl_pb2.TRAINING,
+                    elasticdl_pb2.EVALUATION,
+                ]:
+                    if (cur_time - start_time) > 3 * avg_time[task.type]:
+                        logger.info(
+                            "worker %d timeout, relaunch it" % worker_id
+                        )
+                        self.recover_tasks(worker_id)
+                        # TODO: save worker logs before remove it
+                        self._invoke_task_timeout_callback(worker_id)
+                        break
+            time.sleep(30)
+
+    def _get_average_task_completed_time(self):
```

Comment:
We can add a description that the return value is a dict containing the average time for TRAINING and EVALUATION tasks.